enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,432 @@
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+1. ~~How to handle upgrades?~~
+2. Which release should this change target?
+
+## Summary
+
+The `cluster-baremetal-operator` component will provide bare metal
+machine management capabilities needed for the Machine API provider on
+the `baremetal` platform. There is no equivalent for this component on
+other platforms.
+
+This functionality is currently provided by `machine-api-operator`
+directly via `baremetal-operator` from the Metal3 project, and related
+components. This situation requires the Machine API Operator to have
+significant bare metal specific knowledge.
+
+This proposal outlines a plan for a new bare metal specific component
+as a fully-fledged Second Level Operator under the management of the
+Cluster Version Operator.
+
+The new `cluster-baremetal-operator` will initially merely adopt the
+existing bare metal specific components - including
+`baremetal-operator` from `machine-api-operator`. However, we expect
+that additional bare metal specific functionality will be added to
+this new operator over time.
+
+## Motivation
+
+In order to bring up a cluster using the baremetal platform:
+
+1. The installer needs to capture bare metal host information and
+   provisioning configuration for later use.
+2. Something needs to install the CRDs for these bare metal specific
+   resources created by the installer.
+3. Something needs to launch a controller for these resources.
+
+Currently (1) is achieved by the installer creating:
+
+* A Provisioning resource manifest
+* Manifests for BareMetalHost resources and their associated secrets
+
+and these manifests are applied by the cluster-bootstrap component
+towards the end of the cluster bootstrapping process.
+
+This resource creation step does not succeed until the step (2)
+completes - i.e. the relevant CRDs are applied - and this is currently
+done by the Cluster Version Operator (CVO) as it applies the manifests
+for the MAO.
+
+Finally, (3) happens when the MAO detects that it is running on the
+`baremetal` infrastructure platform and instantiates a relatively
+complex Deployment including the BareMetalHost controller and various
+containers running other components, all from the Metal3 project. The
+configuration of this deployment is driven by the Provisioning
+resource created by the installer in (1).
+
+There are two problems emerging with this design:
+
+* A sense that the MAO is responsible for a significant amount of bare
+  metal specific tasks that should be declared outside of the scope of
+  the MAO, particularly since it does not have equivalent
+  responsibilities on any other platform.
```

Comment:
Does this also apply to vSphere? Is [the vSphere actuator](https://github.com/openshift/machine-api-operator/tree/master/pkg/controller/vsphere) going to be moved as part of this work?

Reply:
Looking back to a discussion elsewhere in February, this seems like a continuation of a question you asked there, so I'm going to include that here.

Doug:

>  I would like a technical explanation of why a new operator is the right approach. Why is the 
> operator responsible for standing up the controller for the machine API not the right place 
> for doing the work to stand up the controller for the machine API for bare metal?

David Eads @deads2k :

> The machine-api-operator is responsible for making a call to the cloud provider API to get a 
> machine provisioned, but it is not responsible for running the components that cloud 
> provider API requires in order to function.  My understanding of the metal3 additions to the 
> machine-api-operator are that it has two parts
>
> 1. An actuator that knows how to make a request for a machine.  This actuator clearly 
> belongs in the machine-api-operator.
> 2. A metal3 deployment which contains an ipa-downloader, a machine-os-downloader, an 
> ip-setter, a maria db, an ironic conductor, ironic api, and an ironic inspector.  And I have a 
> memory of this even having a way to provide DHCP for the machines provided.  This 
> doesn't appear to belong in the machine-api-operator.
>
> The machine-api-operator is a client to a cloud provider API which provisions machines.  
> The actuator for vsphere doesn't include infrastructure for vsphere to be able launch its 
> machines, only the calls to the API to create them.  We would expect metal3 to do the 
> same.

Reply:
Davids explanation is pretty reasonable I think, but Doug's follow on question is equally reasonable.

That is ... why is the vsphere actuator the only actuator codebase that actually lives in the machine-api-operator codebase?

Or ... why does the `machine-controller-manager` binary for vsphere come from the machine-api-operator image? What's so special about vsphere?

I'm tempted to guess that the only reason it's special is because the authors of the vsphere provider are also the maintainers of the machine-api-operator repo? But it'd be great to get that background from @enxebre

Reply:
Just for reference, I don't see any of the rationale for that in e.g. openshift/machine-api-operator#447 or openshift/enhancements#154

Reply:
Ah, so the trend would actually be in the opposite direction? All of the cluster-api-provider repos should be moving their code into the MAO repo? Presumably including cluster-api-provider-baremetal?

Reply:
I guess so, though I'll defer to @enxebre as I joined the team after this was done so I'm not entirely sure what the plan was

Reply:
The machine API operator is intentionally simple and fairly rudimentary by design.
It literally just enforces the lifecycle of a kubernetes deployment resource.
This deployment has the same orthogonal structure everywhere. It know nothing about clouds.

Now, where it lives the code base for the containers which are run by that deployment has zero impact on the design described above.

>I'm tempted to guess that the only reason it's special is because the authors of the vsphere provider are also the maintainers of the machine-api-operator repo?

Yes. The home for that code base responds to factors totally irrelevant for mao design. These factors are variable in time and context.
You'll see aws, gcp and azure in individual repos because that made sense at the time they were created. They were forks mimicking and upstream code organisation.
vSphere was created in a completely different context. Have it in its individual repo would have only drawbacks but no benefits. Therefore we choose to put the code base in mao repo purely to enable a healthier development workflow (e.g remove dependency revendoring hassle, on the fly CI).

The metal3 bare metal or the openstack provider could well live in the mao repo and that would make no difference at all on the mao design. It is just natural and convenient for them to be individual repos to enable a more effective development workflow given our teams organisation.

Now managing the lifecycle of the underlying cloud required for the machine API to work, designing and understanding how to fetch the required external config, give semantics to possible failure scenarios and communicate those appropriately, that's beyond mao design scope and that's only relevant in a bare metal environment. This is the knowledge that I think is reasonable to embed in an ad-hoc operator and it would favour a more sustainable component ecosystem with clear responsibility boundaries. mao assumes this underlying "cloud" exists.